nixos/xserver: add option to install custom xkb layouts

rnhmjoj opened the pull request:
###### Motivation for this change
I have found no way to add a custom keyboard layout to xkb declaratively from nixos configuration.
I wrote this module based on a blog [post](http://stesie.github.io/2018/09/nixos-custom-keyboard-layout-revisited) and it's working as intended so far. See also #31138.

The usage is as follows:
``` 
  services.xserver = {  
    layout = "rn";
    extraLayouts.rn = {
      description = "US with alt-gr, greek, math and other simbols";
      languages = [ "eng" "ita" ];
      filepath = /home/rnhmjoj/.config/xkb/symbols/rn;
    };
  };
```

It works by patching various packages so I'm not sure if this is acceptable.

###### Things done
- [x] Test in a NixOS VM
- [x] Test more than one layout
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

rnhmjoj committed:
nixos/xserver: add option to install custom xkb layouts

peterhoeg commented:
This is great - I'm (hopefully) going to try this out over the weekend. Instead of overriding the various packages, can't we use a wrapper derivation instead that consists of vanilla upstream + whatever additional layouts you want?

rnhmjoj commented:
Update? I don't think it's possibile to do that because the file `base.xml` and others must be modified to account for the added layouts

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,150 @@
+
+  config = mkIf (layouts != { }) {
+
+    nixpkgs.config.packageOverrides = super: {
```
This should use overlays instead.

rnhmjoj replied:
I'm not familiar with overlays, i'll look into it.

rnhmjoj replied:
Ok, I think I did it.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,150 @@
+
+      xkbvalidate = super.xkbvalidate.override {
+        libxkbcommon = super.libxkbcommon.override {
+          xkeyboard_config = xkb_patched;
```
Having to override everything that uses xkb manually isn't very nice. By changing xorg's `xkeyboard_config` attribute you should be able to change those all with one setting (unless I'm missing something).

rnhmjoj replied:
In theory yes, but that would require to recompile a huge amount of software. This trick minimizes that to only a few packages.

Infinisil replied:
Ah darn :/

rnhmjoj commented:
I tested it again with overlays and seems ok. I also added a comment to explain the situation.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,154 @@
+        xorgserver = super.xorg.xorgserver.overrideAttrs (old: {
+          configureFlags = old.configureFlags ++ [
+            "--with-xkb-bin-directory=${self.xorg.xkbcomp}/bin"
+            "--with-xkb-path=${xkb_patched}/share/X11/xkb"
```
Oh actually, doesn't the `services.xserver.xkbDir` option do the same, but without having to recompile?

rnhmjoj replied:
Yeah, but it doesn't really work: see #31138.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,154 @@
+      cp rules/base.xml.in rules/base.evdev.xml.in
+  '';
+
+  xkb_patched = pkgs.xorg.xkeyboardconfig.overrideAttrs (old: {
```
I just noticed this, you're defining `xkb_patched` in terms of `pkgs`, but then use `xkb_patched` in an overlay below, but `pkgs` needs to evaluate all overlays to evaluate, so I think this means it needs to evaluate nixpkgs twice to get the results. You should define `xkb_patched` within the overlay (in terms of `self`) instead

Infinisil replied:
Wouldn't hurt to also export the patched version as an attribute in the overlay.

rnhmjoj replied:
Uhm, does `patchIn`, which needs `pkgs.ed`, has to be moved in the overlay as well?

rnhmjoj replied:
Ok, done.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,154 @@
+        '';
+      };
+
+      filepath = mkOption {
```
One last nitpick (sorry, last one for sure this time, from me at least): I'd prefer `layoutFile` over `filepath`, the former is more in the style of other NixOS options.

rnhmjoj replied:
Ok, sure.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,154 @@
+
+      filepath = mkOption {
+        type = types.path;
+        description = "The path to the layout file.";
```
Oh actually one more thing: you should add some info on the format of this file, maybe link to https://wiki.archlinux.org/index.php/X_KeyBoard_extension#Basic_examples

rnhmjoj replied:
I linked the official documentation in the description of `extraLayouts`. I'm adding this as well as it provides more examples.

Infinisil commented:
LGTM, but I'd like another pair of eyes to take a look too, the way only certain things get patched is very unorthodox

rnhmjoj commented:
Thank you for the review! I'm aware of that, unfortunately I don't know of any other way.

rnhmjoj commented:
From the blog post, besides xkbcomp and xbkmap:

- xkbvalidate: needed to make the keyboard layout test pass
- xserver: otherwise the xkbcomp will not find the layout even if setxkbmap accepts it.

Infinisil commented:
Okay so I tried this PR out myself, with the following config:

```nix
{
  services.xserver = {
    exportConfiguration = true;
    extraLayouts.cdvp = {
      description = "Custom Programmer Dvorak";
      languages = [ "eng" ];
      layoutFile = builtins.toFile "layout.xkb" ''
        xkb_keymap {
	      xkb_keycodes  { include "evdev+aliases(qwerty)"  };
	      xkb_types     { include "complete"	};
	      xkb_compat    { include "complete"	};
	      xkb_symbols   {
		    include "pc+us(dvp)+inet(evdev)+capslock(backspace)"
		    key <LSGT> { [ Control_L ] };
		    key <RWIN> { [ Break ] };
		    modifier_map Control { <LSGT> };
	      };
	      xkb_geometry  { include "pc(pc105)"	};
       };
     '';
   };
 };
```

But no luck, after a bunch of compilations (and even restarting my display manager), it's as if nothing changed, there's no `cdvp` file in `/etc/X11/xkb/symbols`, as I'd have expected it (`exportConfiguration` should make that work), and `setxkbsymbols cdvp` says `Error loading new keyboard description` (which it says for all layouts that aren't known).

What am I missing here?

Infinisil commented:
I also tried

```nix
{
  layoutFile = builtins.toFile "layout" ''
    xkb_symbols {
      include "pc+us(dvp)+inet(evdev)+capslock(backspace)"
      key <LSGT> { [ Control_L ] };
      key <RWIN> { [ Break ] };
      modifier_map Control { <LSGT> };
    };
  '';
}
```

but no luck either, same result

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,156 @@
+      layoutFile = mkOption {
+        type = types.path;
+        description = "The path to the layout file.";
+      };
```
This indicates strongly that this option should have a good example and a better description with some links to docs or so.

Infinisil replied:
And @laMudri pointed out that if it's such a symbols file, this option should be called `symbolsFile` instead. I didn't realize this earlier.

Infinisil replied:
Also, the build should probably fail somehow if the file given has the wrong format

laMudri replied:
More thoroughly, the motivation for including the word `symbols` is:
1) Future work might make the files in `compat/`, `geometry/`, `keycodes/`, and `types/` customisable in this options block. So the file in `symbols/` is not necessarily the whole layout. (I'm not sure what these files do, but in principle at least they might be customisable. Who knows what the future might bring?)
2) `symbols` is xkb jargon, which might be helpful to see for a user who is trying to work out what to put in this file. As @Infinisil's test case showed, it's not obvious that `layoutFile` is just for symbols, and not for other stuff too.

rnhmjoj replied:
I agree. Including other sections shouldn't be much different, I'll give it a try.
Anyway, I updated the description: better?

rnhmjoj commented:
So, I fixed the issue left when renaming the attributes.
I tested adding more than one layout and it works as expected.
I also added options to install all possible xkb components (compat, geometry, ...) but I haven't tested it properly. I doesn't break the existing functionality, though.

mix-yasuno commented:
Will this PR ever be merged? I tried this PR and worked as expected.

Infinisil commented:
Well, I don't think there's any point in including compat/geometry/keycodes/types when it's untested and probably nobody will use it. @laMudri only mentioned it as potential future work, nothing that needs to be done now. So I'd actually like for those to not be part of this PR.

Two other things that need to be changed:
- If possible, make the build fail if you provide an invalid layout, when I tested it with an invalid layout it just built it without problem.
- Add at least a good and extensive example for the layout file option, this will be very useful for people wanting to configure that option.

mix-yasuno commented:
This PR does not work anymore on `nixos-unstable` channel because `xkeyboard-config` is updated.
Can be fixed by renaming `rules/base.xml.in` to `rules/base.xml` and removing `cp rules/base.xml.in rules/base.evdev.xml.in`.

rnhmjoj commented:
@mix-yasuno: Thank you. I'm now working on documenting the whole thing in the manual.
I'll probably write a NixOS test to pick up this sorts of issues in the future.

rnhmjoj commented:
So, I carefully tested the components of an xkb keymap and everything does work. I also documented in the NixOS manual how to use `xserver.extraLayouts` with two examples and linked the usual references.

There is a minor inconvenience: once a custom keycodes, geometry, types or compat file is installed it's currently not possibile to enable it from configuration.nix and the default xkb rules (obviously) will not select it; `setxkbmap -keycodes` etc. does work, though.
Unfortunately the Xorg server does not support `-xkbmap` anymore. Looking for some alternative I have
found in the [documentation](https://www.x.org/releases/X11R7.5/doc/input/XKB-Config.html) options like `XkbKeycodes`, that I tried, but they don't seem to work.

Infinisil commented on nixos/doc/manual/configuration/x-windows.xml:
```diff
@@ -157,4 +159,123 @@
+  <para>
+   The function <literal>pkgs.writeText &lt;filename&gt; &lt;content&gt;
+   </literal> can be useful if you prefer to keep the layout definitions
+   inside the NixOS configuraiton.
```
configura**ti**on

rnhmjoj replied:
Thank you

Infinisil commented on nixos/doc/manual/configuration/x-windows.xml:
```diff
@@ -125,10 +125,12 @@
+   For recent AMD GPUs you most likely want to keep either the defaults
    or <literal>"amdgpu"</literal> (both free).
   </para>
+  </note>
```
Accidentally included this unrelated change

rnhmjoj replied:
I actually did this on purpose: <note> is used inconsistently in this section of the manual. Is it ok to include it in this change?

Infinisil replied:
Ah yeah sounds good, but a separate commit for this would be nice

rnhmjoj replied:
Ok, done.

Infinisil commented on pkgs/development/python-modules/azure-cli-core/default.nix:
```diff
@@ -33,11 +33,11 @@
   src = fetchPypi {
     inherit pname version;
-    sha256 = "0fp6b2x1l9bg07pca7asm80rnjlc4kkm061s3nrb55yj6awsnim5";
+    sha256 = "797c4fab2285aa2ac316daf692c1e8f6b14186c059805e8a57332b4d83d7bb23";
```
This change is also accidentally included

rnhmjoj replied:
This sometimes happens when I do an interactive rebase and go a commit too far. I don't know how to make github not pick these up, besides rebasing again.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,232 @@
+  patchIn = pkgs: name: layout: with layout;
+  ''
+      # install layout files
+      ${optionalString (compatFile   != null) "cp '${compatFile}'   'compat/${name}'"}
```
And if you have the motivation, after another thought it would make a whole lot more sense for this patching to happen in the xkeyboardconfig derivation directly via an `extraSymbolFiles` argument you can `.override`.

rnhmjoj replied:
It sounds like a good idea but the expression for xkeyboardconfig is generated, I don't know if it's possible to add new arguments.

Infinisil replied:
How about adding a separate derivation `xkeyboardconfig_customizable` which directly takes these `extra*` arguments in https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/x11/xorg/overrides.nix

Something like
```nix
{
  xkeyboardconfig_customizable = { extraLayouts ? {} }:
    self.xkeyboardconfig.overrideAttrs ...;
}
```

rnhmjoj replied:
@infinisil do you another way to add options to an existing package? This does work but `xkeyboardconfig_customizable` will be a function instead of a derivation.

Infinisil replied:
I think this is fine, it's also how other attributes like this work

rnhmjoj replied:
Ok, then I think this is quite ready.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,232 @@
+      description = ''
+        Extra custom layouts that will be included in the xkb configuration.
+        Information on how to create a new layout can be found here:
+        https://www.x.org/releases/current/doc/xorg-docs/input/XKB-Enhancing.html#Defining_New_Layouts
```
You need to use the xml link syntax here to have it be clickable in the manual

rnhmjoj replied:
Should be fixed now.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,232 @@
+      };
+
+      xkbvalidate = super.xkbvalidate.override {
+        libxkbcommon = super.libxkbcommon.override {
```
I believe this should be `self.libxkbcommon`

rnhmjoj replied:
It doesn't seem to make a difference, though: the build produces the exact same hash.

Infinisil replied:
Yeah, but it would make a difference if a different overlay after this one overwrote `libxkbcommon`. With `super` it would not apply, with `self` it would.

infinisil reviewed (changes requested):
Nice works, I'm liking the docs for this!

However you haven't addressed what I mentioned in https://github.com/NixOS/nixpkgs/pull/47764#issuecomment-477968630. In particular I still don't think it's a good idea to support file types other than `symbols`. Those won't be used by anybody (or do you know anybody that does?), so it's essentially dead code that only bloats the module and option listing. If you insist on keeping those then sure, but I'd prefer to not have them included.

rnhmjoj commented:
@Infinisil Adressing your points:

> If possible, make the build fail if you provide an invalid layout, when I tested it with an invalid layout it just built it without problem.

If the layout is installed and loaded with `xserver.layout`, the `xkb-validate` tool (thanks to aszlig) will check it by running xkbcomp. So, if the layout has a syntax error, the configuration will not build. It will even show the line/character where the error occured.

> Add at least a good and extensive example for the layout file option, this will be very useful for people wanting to configure that option.

I think I covered pretty much the main use cases in the manual: other examples would be minor variations and quite pointless.

> In particular I still don't think it's a good idea to support file types other than symbols. Those won't be used by anybody (or do you know anybody that does?), so it's essentially dead code that only bloats the module and option listing. If you insist on keeping those then sure, but I'd prefer to not have them included.

Well, yes, it's probably pretty uncommon to define new keycodes, key types, ... but while researching for this patch I've learned there definitely is a niche around this: like designing layouts for new languages or custom keyboards, I have found several applications that render the physical layouts described by xkb_geometry, etc. NixOS would be the only OS around to make this kind of work really easy, also calling it extraLayouts without being able to really define a full xkb layout would not be right.

rnhmjoj committed:
docs/xserver: use <note> tag for notes

rnhmjoj committed:
docs/xserver: document xserver.extraLayouts

Infinisil commented:
You should also set `services.xserver.xkbDir` to the newly patched one

I guess it's fine to include these configurations then, though I'm still not a fan of it :)

Infinisil commented:
You should also set `services.xserver.xkbDir` to the newly patched one

I guess it's fine to include these configurations then, though I'm still not a fan of it :)

Infinisil commented:
You should also set `services.xserver.xkbDir` to the newly patched one

I guess it's fine to include these configurations then, though I'm still not a fan of it :)

Infinisil commented:
You should also set `services.xserver.xkbDir` to the newly patched one

I guess it's fine to include these configurations then, though I'm still not a fan of it :)

Infinisil commented:
(Apparently my comment was recorded even though GitHub was down)

rnhmjoj commented:
> You should also set services.xserver.xkbDir to the newly patched one

Done.

rnhmjoj commented:
I also moved the patchable `xkeyboardconfig` to the xorg packages set.

Infinisil commented on nixos/modules/services/x11/extra-layouts.nix:
```diff
@@ -0,0 +1,165 @@
+          The path to the xkb symbols file.
+          This is the most important file: it defines which symbol or action
+          maps to each key and must contain a
+          <literal>xkb_symbols { ... }</literal> block.
```
While testing this I noticed that this isn't correct: You need to give a named block like `xkb_symbols "layout_name" { ... }` to not have it fail. Probably the same for the other files.

rnhmjoj replied:
Oh, yes. I'll fix it.

Infinisil commented:
@GrahamcOfBorg eval

infinisil reviewed (approved):
Looks good to me!

rnhmjoj commented:
@Infinisil Thank you for all the help!

mix-yasuno commented:
Yes, finally! Thank you guys!

dasj19 commented:
Hello, I tried to use this to add a system-wide custom keyboard layout. But this does not seem to work in gnome.

I have the following in my configuration.nix:
```
  services.xserver.layout = "esrodk,es,dk,ro";
  services.xserver.extraLayouts.esrodk = {
    description = "Spanish layout with romanian and danish diacritics";
    languages = ["spa"];
    symbolsFile = /etc/nixos/esrodk;
  };
```

When I build and log in gnome the new layout shows up as a language instead of a variant of a language layout.
I come from debian and there I applied the following patch:
```
diff -ur /usr/share/X11/xkb/symbols/es /usr/share/X11/xkb/symbols/es
--- /usr/share/X11/xkb/symbols/es	2019-02-10 22:47:30.962363954 +0100
+++ /usr/share/X11/xkb/symbols/es	2019-02-10 23:03:05.664200582 +0100
@@ -264,3 +264,29 @@
 	xkb_symbols "sun_type6" {
 	include "sun_vndr/es(sun_type6)"
 };
+
+partial alphanumeric_keys
+xkb_symbols "esrodk" {
+
+    include "es(basic)"
+
+    name[Group1]="Spanish +RO/DK";
+
+    key <AE10>	{ [         0,      equal,   0x00002260,       degree ]	};
+    key <AE11>	{ [apostrophe, question, doublelowquotemark, question ]	};
+    key <AE12>	{ [exclamdown, questiondown, rightdoublequotemark, exclam ]	};
+
+    key <AD01>  { [         q,          Q,  acircumflex,  Acircumflex ] };
+    key <AD02>	{ [         w,          W,        aring,	Aring ]	};
+    key <AD05>  { [         t,          T,    0x100021b,    0x100021a ] };
+    key <AD06>	{ [         y,          Y,    leftarrow,   rightarrow ]	};
+    key <AD07>	{ [         u,          U,    uparrow,      downarrow ]	};
+    key <AD08>  { [         i,          I,  icircumflex,  Icircumflex ] };
+
+    key <AC01>  { [         a,          A,       abreve,       Abreve ] };
+    key <AC02>  { [         s,          S,    0x1000219,    0x1000218 ] };
+
+    key <LSGT>	{ [      less,    greater,    	     ae,	   AE ]	};
+
+    include "level3(ralt_switch)"
+};
diff -ur /usr/share/X11/xkb/rules/evdev.lst /usr/share/X11/xkb/rules/evdev.lst
--- /usr/share/X11/xkb/rules/evdev.lst	2019-02-10 22:46:55.358727634 +0100
+++ /usr/share/X11/xkb/rules/evdev.lst	2019-02-10 23:02:15.597083721 +0100
@@ -650,6 +650,7 @@
   ast             es: Asturian (Spain, with bottom-dot H and bottom-dot L)
   cat             es: Catalan (Spain, with middle-dot L)
   mac             es: Spanish (Macintosh)
+  esrodk            es: Spanish (+RO/DK)
   nodeadkeys      se: Swedish (no dead keys)
   dvorak          se: Swedish (Dvorak)
   rus             se: Russian (Sweden, phonetic)
diff -ur /usr/share/X11/xkb/rules/evdev.xml /usr/share/X11/xkb/rules/evdev.xml
--- /usr/share/X11/xkb/rules/evdev.xml	2019-02-10 22:47:19.018488400 +0100
+++ /usr/share/X11/xkb/rules/evdev.xml	2019-02-10 23:03:57.375281353 +0100
@@ -4930,6 +4930,12 @@
             <description>Spanish (Macintosh)</description>
           </configItem>
         </variant>
+	<variant>
+          <configItem>
+            <name>esrodk</name>
+            <description>Spanish (+RO/DK)</description>
+          </configItem>
+        </variant>
       </variantList>
     </layout>
     <layout>
@@ -7388,4 +7394,4 @@
       </option>
     </group>
   </optionList>
-</xkbConfigRegistry>
\ Nici un element de linie nouă la sfârşitul fişierului
+</xkbConfigRegistry>
```
Now when I look at the files that get builded from configuration.nix ... in the evdev.lst there is an entry:
```
! model
..........
  teck227         Truly Ergonomic Computer Keyboard Model 227 (Wide Alt keys)
  teck229         Truly Ergonomic Computer Keyboard Model 229 (Standard sized Alt keys, additional Super and Menu key)
  esrodk 
! layout
...........
  md              Moldavian
  id              Indonesian (Jawi)
  my              Malay (Jawi, Arabic Keyboard)
  esrodk 
```
Note that esrodk has no human-friendly name.. this might be the reason it does not show up in gnome-control-center under region settings.

rnhmjoj commented:
> But this does not seem to work in gnome.

I'm not a gnome use so I've never tested this. What does `localectl` says? Is the layout selected?
You could also set `i18n.consoleUseXkbConfig = true;` and see if it works in the linux vt;

> When I build and log in gnome the new layout shows up as a language instead of a variant of a language layout.

This is intended: this options installs a layout by placing a new file in the xkb source tree.
To make a layout variant It would have to edit an existing xkb: It can be done but it's more complicated. Maybe in the future...

> Note that esrodk has no human-friendly name.. this might be the reason it has issues with gnome-control-center region settings.

The description should be in there... this maybe be a bug.

dasj19 commented:
The result of localectl... 
```
[root@nixos:~]# localectl
   System Locale: LANG=ro_RO.UTF-8
       VC Keymap: es
      X11 Layout: esrodk,es,dk,ro
       X11 Model: pc104
     X11 Options: terminate:ctrl_alt_bksp

```

I'll get back when I build with `i18n.consoleUseXkbConfig = true;`

Another thing... it is really needed to build all the gnome packages from source when using `services.xserver.extraLayouts` ?

dasj19 commented:
Hello I can confirm that with the following:
```
  # Enable the X11 windowing system.
  services.xserver.enable = true;
  services.xserver.layout = "esrodk,es,dk,ro";
  services.xserver.xkbModel = "esrodk";
  services.xserver.extraLayouts.esrodk = {
    description = "Spanish layout with romanian and danish diacritics";
    languages = ["spa"];
    symbolsFile = /etc/nixos/esrodk;
  };

  # Select internationalisation properties.
  i18n = {
    consoleFont = "Lat2-Terminus16";
    consoleUseXkbConfig = true;
    defaultLocale = "ro_RO.UTF-8";
    supportedLocales = [ "en_US.UTF-8/UTF-8" "ro_RO.UTF-8/UTF-8" "da_DK.UTF-8/UTF-8" "es_ES.UTF-8/UTF-8" ];
  };
```
configuration in configuration.nix I can use "esrodk" as default in a tty terminal... but still does not work in gnome.
And I believe the reason for this is the missing human-friendly name in the xkeyboard-config-2.27/share/X11/xkb/rules/evdev.lst ... as I only have a few days experience with nixos I don't know how to test my theory/debug further.

localectl ouputs the following now:
```
System Locale: LANG=ro_RO.UTF-8
       VC Keymap: /nix/store/k43rdd6m1c1b713xisynkqmqi45j18km-xkb-console-keymap
      X11 Layout: esrodk,es,dk,ro
       X11 Model: esrodk
     X11 Options: terminate:ctrl_alt_bksp
```

rnhmjoj commented:
> Another thing... it is really needed to build all the gnome packages from source when using services.xserver.extraLayouts ?

We are already trying to reduce the number of rebuilds by not overriding xkb directly, however xorgserver, setxkbmap, xkbcomp, ckbcomp and xkbvalidate *must* be overriden.

It seems the problem is gdm and mutter depends on xorgserver.
Perhaps we could pass xkbcomp to the X server without reconfiguring the build but it seems unlikely.

> Update: I fixed the human-friendly name issue in this commit... dasj19@87ec211

Oh, right: that broke when xkb dropped intltool. Could you make a PR?

> however I'm still trying to get it working in gnome.

I'll try it too as soon as I have some spare time; I should also look into other DEs as well. I know it, at least, works in XFCE.

dasj19 commented:
Some developments...
1. I opened the above PR
2. I asked over at gnome about a related issue which seems to be a side-effect : https://gitlab.gnome.org/GNOME/libgnomekbd/issues/3
3. I noticed that I can use my new keyboard layout in gnome if I do `setxkbmap esrodk`
4. The output of `setxkbmap -print -verbose 10` is:
```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /nix/store/l246g9yz3y85qaiqh3y510qc0yq77j3v-setxkbmap-1.3.2/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      es
layout:     es,dk,ro,us
variant:    ,,,
options:    terminate:ctrl_alt_bksp
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+es+dk:2+ro:3+us:4+inet(evdev)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+es+dk:2+ro:3+us:4+inet(evdev)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc104)"	};
};
```
and after `setxkbmap esrodk`
```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /nix/store/l246g9yz3y85qaiqh3y510qc0yq77j3v-setxkbmap-1.3.2/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      es
layout:     esrodk
options:    terminate:ctrl_alt_bksp
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+esrodk+inet(evdev)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+esrodk+inet(evdev)+terminate(ctrl_alt_bksp)"};
	xkb_geometry  { include "pc(pc104)"	};
};
```
Note: the layout part of the 2 outputs.
I do not know if my findings are of any help... but I guess they are a step in the right direction

rnhmjoj commented:
Did you try setting only one layout? I mean like `services.xserver.layout = "esrodk";`

dasj19 commented:
I just tried... It looks like it uses the US layout ...

This is what i have in configuration.nix:
```
  # Enable the X11 windowing system.
  services.xserver.enable = true;
  services.xserver.layout = "esrodk";
  services.xserver.xkbModel = "esrodk";
  services.xserver.extraLayouts.esrodk = {
    description = "Spanish layout with romanian and danish diacritics";
    languages = ["spa"];
    symbolsFile = /etc/nixos/esrodk;
  };

  # Select internationalisation properties.
  i18n = {
    consoleFont = "Lat2-Terminus16";
    consoleUseXkbConfig = true;
    defaultLocale = "ro_RO.UTF-8";
    supportedLocales = [ "en_US.UTF-8/UTF-8" "ro_RO.UTF-8/UTF-8" "da_DK.UTF-8/UTF-8" "es_ES.UTF-8/UTF-8" ];
  };
```

After rebooting GDM uses US layout, same as the rest of Gnome, while TTY uses "esrodk".

Some more info:
Output of `setxkbmap -print -verbose 10`
```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /nix/store/lx0czczcaribbiiy6frsrv4g70vb6ic7-setxkbmap-1.3.1/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      esrodk
layout:     us,us
variant:    ,
options:    terminate:ctrl_alt_bksp
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+us:2+inet(evdev)+terminate(ctrl_alt_bksp)
geometry:   pc(pc104)
xkb_keymap {
	xkb_keycodes  { include "evdev+aliases(qwerty)"	};
	xkb_types     { include "complete"	};
	xkb_compat    { include "complete"	};
	xkb_symbols   { include "pc+us+us:2+inet(evdev)+terminate(ctrl_alt_bksp)"	};
	xkb_geometry  { include "pc(pc104)"	};
};
```

Output of localectl:
```
   System Locale: LANG=ro_RO.UTF-8
       VC Keymap: /nix/store/phd5cpzk3rrbn6px97pj89nn46i4yign-xkb-console-keymap
      X11 Layout: esrodk
       X11 Model: esrodk
     X11 Options: terminate:ctrl_alt_bksp
```

Next thing I'll try is to remove `"en_US.UTF-8/UTF-8"`  from `i18n.supportedLocales`

rnhmjoj commented:
@dasj19 So, I have tested KDE Plasma, XFCE and GNOME. It seems to be a problem only in GNOME. I couldn't figure out much besides wht you already found.
Could you open a issue and continue the discussion there?

jtojnar commented:
The GNOME issue was fixed in https://github.com/NixOS/nixpkgs/pull/76591

attila-lendvai commented:
@rnhmjoj thanks a whole lot for this!

a question: is it expected that a whole lot of packages are compiled locally on 20.09 stable? i mean even stuff like the x server and gnome?

```
  services.xserver.layout = "us";
  services.xserver.extraLayouts.hu_attila = {
    description = "Attila's modified hu layout";
    languages   = [ "hun" ];
    symbolsFile = ./xkb/hu_attila;
  };
```

dylanowen commented:
I'm pretty new to Nix and xkb but I've been trying to setup a keyboard layout using custom types. To actually use xkb types applied using this method it seems like we would need to update the `types/complete` file to reference the newly created `types/${name}` file? Is there a better way to do this? If not I can open a PR with the necessary changes once I get a chance to test them.

rnhmjoj commented:
I don't know how `types` files work, is it always necessary to update `types/complete` after a new one is added?
I see there are a number of types files that are not references by `types/complete`.

Anyway, If you need to replace an existing file in xkb, you could try to define another layout like:
```
extraLayouts.complete.typesFile = ...;
```
This should override `types/complete`. Otherwise, you have to extend the script in `pkgs/servers/x11/xorg/overrides.nix:483`.

dylanowen commented:
@rnhmjoj Thanks for the quick response. After looking at your comment I did some more research and I think your work on the rules file should cover what I was imagining. I'll give it a try when I'm back to a computer.

attila-lendvai commented:
the loading of user specific xkb files should be ready according to this: https://who-t.blogspot.com/2020/09/user-specific-xkb-configuration-putting.html

not sure when these changes get/got into NixOS.

dylanowen commented:
@rnhmjoj I haven't been able to figure out how types are attached besides [this post.](https://unix.stackexchange.com/questions/240566/what-is-wrong-with-these-xkb-rules-files/355404#355404) But I was attempting to try your suggestion of overwriting the `complete` types file using this
```nix
services.xserver.extraLayouts.complete = {
   description = "Complete types override";
   languages = [ "eng" ];
   typesFile = ./keyboard/complete.xkb;
  };
```
and I ran into this error:
```shell
make[2]: Nothing to be done for 'install-exec-am'.
 /nix/store/1c1r48qa0m23vr9jy8sm0dc04vv14dak-coreutils-8.32/bin/mkdir -p '/nix/store/xlqwv7waszd0f8xk651jfrml4fw2301b-xkeyboard-config-2.31/share/X11/xkb/compat'
 /nix/store/1c1r48qa0m23vr9jy8sm0dc04vv14dak-coreutils-8.32/bin/install -c -m 644 complete accessx basic caps complete iso9995 japan ledcaps ledcompose lednum ledscroll level5 misc mousekeys olpc pc pc98 xfree86 xtest README '/nix/store/xlqwv7waszd0f8xk651jfrml4fw2301b-xkeyboard-config-2.31/share/X11/xkb/compat'
/nix/store/1c1r48qa0m23vr9jy8sm0dc04vv14dak-coreutils-8.32/bin/install: will not overwrite just-created '/nix/store/xlqwv7waszd0f8xk651jfrml4fw2301b-xkeyboard-config-2.31/share/X11/xkb/compat/complete' with 'complete'
```
Should we reverse the check in this line https://github.com/NixOS/nixpkgs/blob/a78925d568e884da7e7812bed09e02c750e8d3b0/pkgs/servers/x11/xorg/overrides.nix#L506 so that we're adding our types to our Makefiles if they don't already exist? My assumption of what was happening was that the makefile was ending up with duplicates of the `complete` line in the `_DATA` section. Also should we only be adding files to the `_DATA` section if they were configured? I believe right now we'd be putting files in the `_DATA` section that don't exist.
